r2-core/src/main/java/com/linkedin/r2/message/RequestContext.java

```diff
@@ -131,4 +140,204 @@ public String toString()
   {
     return _localAttrs.toString();
   }
+
+  private static class AttrsMap<K, V> implements Map<K, V>, Lock {
+    private final Map<K, V> map;
+    private final Lock lock = new ReentrantLock();
+
+    public AttrsMap() {
+      map = new HashMap<>();
+    }
+
+    public AttrsMap(Map<K, V> other) {
+      map = new HashMap<>(other);
+    }
+
+    /**
+     * Create a clone of another AttrsMap.
+     *
+     * This is more efficient than the generic Map-based constructor for AttrsMap since it synchronizes only once for
+     * the entire copy rather instead of also synchronizing for every element read.
+     *
+     * @param other Another AttrsMap to copy.
+     */
+    public AttrsMap(AttrsMap<K, V> other) {
+      other.lock.lock();
+      try {
+        map = new HashMap<>(other.map);
+      } finally {
+        other.lock.unlock();
+      }
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      lock();
+      try {
+        return map.equals(other);
```

Comment:
Seems like this should be testing to see if the other object is AttrsMap, and locking it if so